contracts/Relay.sol

```diff
@@ -0,0 +1,431 @@
+pragma solidity ^0.5.10;
+
+/** @title BitcoinSPV */
+/** @author Summa (https://summa.one) */
+
+import {SafeMath} from "./SafeMath.sol";
+import {BytesLib} from "./BytesLib.sol";
+import {BTCUtils} from "./BTCUtils.sol";
+import {ValidateSPV} from "./ValidateSPV.sol";
+
+
+interface IRelay {
+    function isMostRecentAncestor(
+        bytes32 _ancestor,
+        bytes32 _left,
+        bytes32 _right,
+        uint256 _limit
+    ) external view returns (bool);
+
+    function findHeight(bytes32 _digest) external view returns (uint256);
+
+    function findAncestor(bytes32 _digest, uint256 _offset) external view returns (bytes32);
+
+    function isAncestor(bytes32 _ancestor, bytes32 _descendant, uint256 _limit) external view returns (bool);
+
+    function heaviestFromAncestor(
+        bytes32 _ancestor,
+        bytes calldata _left,
+        bytes calldata _right
+    ) external view returns (bytes32);
+
+    function addHeaders(bytes calldata _anchor, bytes calldata _headers) external returns (bool);
+
+    function addHeadersWithRetarget(
+        bytes calldata _oldPeriodStart,
+        bytes calldata _oldPeriodEnd,
+        bytes calldata _headers
+    ) external returns (bool);
+
+    function markNewHeaviest(
+        bytes32 _ancestor,
+        bytes calldata _currentBest,
+        bytes calldata _newBest,
+        uint256 _limit
+    ) external returns (bool);
+}
+
+
+contract Relay is IRelay {
+    using SafeMath for uint256;
+    using BytesLib for bytes;
+    using BTCUtils for bytes;
+    using ValidateSPV for bytes;
+
+    event Extension(bytes32 indexed _first, bytes32 indexed _last);
+    event Reorg(bytes32 indexed _from, bytes32 indexed _to, bytes32 indexed _gcd);
+
+    bytes32 public relayGenesis;
+    bytes32 public bestKnownDigest;
+    bytes32 public lastReorgCommonAncestor;
+    mapping (bytes32 => bytes32) internal previousBlock;
+    mapping (bytes32 => uint256) internal blockHeight;
+
+
+    /// @notice                   Gives a starting point for the relay
+    /// @dev                      We don't check this AT ALL really. Don't use relays with bad genesis
+    /// @param  _genesisHeader    The starting header
+    /// @param  _height           The starting height
+    constructor(bytes memory _genesisHeader, uint256 _height, bytes32 _periodStart) public {
+        require(_genesisHeader.length == 80, "Stop being dumb");
+        bytes32 _genesisDigest = _genesisHeader.hash256();
+        relayGenesis = _genesisDigest;
```

Comment:
nit: This doesn't seem useful at all in the context of stateless spv, so consider removing it

Reply:
It's marginally useful for consumers to be able to decide if the relay started from an okay block